### PR TITLE
feature: custom alerting

### PR DESCRIPTION
**Pull request by junotx**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Custom alerting with prometheus style is introduced

### CODE REVIEW

**Comment by zryfish** [comment]
This is a huge PR, is there any issues and design documentation for reviewers? And please add more comments to your code, it's read friendly for others.

**Comment by benjaminhuo** [comment]
Related issue: https://github.com/kubesphere/kubesphere/issues/3065

**Comment by benjaminhuo** [comment]
There will be some changes to the implementation, so please hold on reviewing this PR

**Comment by junotx** [comment]
Now it is ready for review @benjaminhuo

File: pkg/api/customalerting/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+type RuleLevel string
+
+type AlertingRuleQualifier struct {
+	Id     string    `json:"id,omitempty" description:"rule id is mainly for the builtin rules"`
+	Name   string    `json:"name,omitempty" description:"rule name which should be uniq for custom rules in the specified namespace"`
+	Level  RuleLevel `json:"level,omitempty" description:"rule level is only for the custom rules and its value is one of cluster, namespace"`
+	Custom bool      `json:"custom" description:"whether to be a custom rule. The builtin rules are not custom and can only be viewed"`
+}
+
+type AlertingRuleProps struct {
```
Comment (benjaminhuo) [inline pkg/api/customalerting/v1alpha1/types.go:55]: As discussed we can adjust AlertingRuleProps PostableAlertingRule GettableAlertingRule as below:
- AlertingRuleProps => AlertingRule
- Move common properties like Alias and Description to AlertingRule
- Move name and id in AlertingRuleQualifier to AlertingRule
- PostableAlertingRule could be replaced by AlertingRule
- Custom and Level could be removed from AlertingRuleQualifier
- Change Rule *AlertingRuleQualifier in Alerts to RuleId and RuleName

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/kapis/customalerting/v1alpha1/handler.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+	}
+	if rule == nil {
+		ksapi.HandleNotFound(resp, nil, err)
+		return
+	}
+
+	resp.WriteEntity(rule)
+}
+
+func (h *handler) handleListBuiltinSpecifiedRuleAlerts(req *restful.Request, resp *restful.Response) {
```
Comment (benjaminhuo) [inline pkg/kapis/customalerting/v1alpha1/handler.go:269]: handleListBuiltinSpecifiedRuleAlerts 
=>
handleListBuiltinRuleAlerts

File: pkg/kapis/customalerting/v1alpha1/handler.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+	ruleId := req.PathParameter("rule_id")
+
+	alerts, err := h.operator.ListBuiltinSpecifiedRuleAlerts(req.Request.Context(), ruleId)
```
Comment (benjaminhuo) [inline pkg/kapis/customalerting/v1alpha1/handler.go:272]: ListBuiltinSpecifiedRuleAlerts
=>
ListBuiltinRuleAlerts

File: pkg/models/customalerting/customalerting.go [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+
+		ruleClient: ruleClient,
+
+		thanosRuleResourceLabels: make(map[string]string),
+	}
+
+	o.resourceRuleCache = rules.NewRuleCache(o.ruleResourceInformer)
+
+	if option != nil && len(option.ThanosRuleResourceLabels) != 0 {
+		lblStrings := strings.Split(option.ThanosRuleResourceLabels, ",")
```
Comment (benjaminhuo) [inline pkg/models/customalerting/customalerting.go:91]: lblStrings = > labels
lblString => label
lbl => l

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

File: pkg/models/customalerting/rules/utils.go [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+	lbls[LabelKeyInternalRuleName] = epRule.Name
+	lbls[LabelKeyInternalRuleQuery] = query
+	lbls[LabelKeyInternalRuleDuration] = duration
+
+	return prommodel.Fingerprint(prommodel.LabelsToSignature(lbls)).String(), nil
+}
+
+// MixAlertingRules mix rules from prometheusrule custom resources and rules from endpoints.
+// Use rules from prometheusrule custom resources as the main reference.
+func MixAlertingRules(ruleNamespace string, ruleChunk *ResourceRuleChunk, epRuleGroups []*customalerting.RuleGroup,
```
Comment (benjaminhuo) [inline pkg/models/customalerting/rules/utils.go:133]: to be more clear, maybe we can rename MixAlertingRules/MixAlertingRule/mixAlertingRule to GetAlertingRulesStatus/GetAlertingRuleStatus/getAlertingRuleStatus ?

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

**Comment by benjaminhuo** [comment]
This is indeed a PR with lots of efforts, thanks @junotx 
Pls make some minor changes as the comments above, thanks again.

File: pkg/api/customalerting/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+	Health                    string     `json:"health,omitempty" description:"health state of a rule based on the last execution, one of ok, err, unknown"`
+	LastError                 string     `json:"lastError,omitempty" description:"error for the last execution"`
+	EvaluationDurationSeconds float64    `json:"evaluationTime,omitempty" description:"taken seconds for evaluation of query expression"`
+	LastEvaluation            *time.Time `json:"lastEvaluation,omitempty" description:"time for last evaluation of query expression"`
+	Alerts                    []*Alert   `json:"alerts,omitempty" description:"alerts"`
+}
+
+type GettableAlertingRuleList struct {
+	Items []*GettableAlertingRule `json:"items"`
+	Total int                     `json:"total"`
```
[the remark this answers is not in the document]
Comment (junotx) [reply pkg/api/customalerting/v1alpha1/types.go:114]: It is for paging

File: pkg/models/customalerting/customalerting.go (hunk 4 above)
Comment (junotx) [reply pkg/models/customalerting/customalerting.go:91]: These names will remain because a package called labels has been imported

Comment (benjaminhuo) [reply pkg/models/customalerting/customalerting.go:91]: ok

**Review by benjaminhuo — APPROVED** [review]
looks better
Please resolve the conflicts @junotx 

And we can set maxConfigMapDataSize to be larger than 0.3

**Review by benjaminhuo — APPROVED** [review]
/lgtm

**Comment by benjaminhuo** [comment]
The previous alerting uses alerting.kubesphere.io/v1 , @junotx pls considering changing customalerting.kubesphere.io/v1alpha1 to alerting.kubesphere.io/v2alpha1 in a new PR if it's feasible

File: pkg/apiserver/config/config.go [hunk 7]
```diff
@@ -96,6 +97,7 @@ type Config struct {
 	AuditingOptions       *auditingclient.Options                    `json:"auditing,omitempty" yaml:"auditing,omitempty" mapstructure:"auditing"`
 	AlertingOptions       *alerting.Options                          `json:"alerting,omitempty" yaml:"alerting,omitempty" mapstructure:"alerting"`
 	NotificationOptions   *notification.Options                      `json:"notification,omitempty" yaml:"notification,omitempty" mapstructure:"notification"`
+	CustomAlertingOptions *customalerting.Options                    `json:"customalerting,omitempty" yaml:"customalerting,omitempty" mapstructure:"customalerting"`
```
[the remark this answers is not in the document]
Comment (benjaminhuo) [reply pkg/apiserver/config/config.go:101]: the option should be combined with the original alert option when changing customalerting.kubesphere.io/v1alpha1 to alerting.kubesphere.io/v2alpha1. @junotx

[the remark this answers is not in the document]
Comment (junotx) [reply pkg/apiserver/config/config.go:101]: This adjustment was discussed yesterday and will be changed in a new pr

File: pkg/simple/client/alerting/options.go [hunk 8]
```diff
@@ -26,13 +39,37 @@ func NewAlertingOptions() *Options {
...
+func (o *Options) AddFlags(fs *pflag.FlagSet, c *Options) {
+	fs.StringVar(&o.Endpoint, "alerting-server-endpoint", c.Endpoint,
+		"alerting server endpoint for the alerting with v1 version.")
+
+	fs.StringVar(&o.PrometheusEndpoint, "alerting-prometheus-endpoint", c.PrometheusEndpoint,
+		"Prometheus service endpoint from which built-in alerting rules are gotten(for the alerting with v2alpha1 version or higher versions)")
+	fs.StringVar(&o.ThanosRulerEndpoint, "alerting-thanos-ruler-endpoint", c.ThanosRulerEndpoint,
+		"Thanos ruler service endpoint from which custom alerting rules are gotten(for the alerting with v2alpha1 version or higher versions)")
+	fs.StringVar(&o.ThanosRuleResourceLabels, "alerting-thanos-rule-resource-labels", c.ThanosRuleResourceLabels,
+		"The labels will be added to prometheusrule custom resources to be selected by thanos ruler. eg: thanosruler=thanos-ruler,role=custom-alerting-rules (for the alerting with v2alpha1 version or higher versions)")
```
Comment (benjaminhuo) [inline pkg/simple/client/alerting/options.go:74]: The labels will be added to prometheusrule custom resources to be selected by thanos ruler =>
Labels used by Thanos Ruler to select PrometheusRule custom resources

File: pkg/simple/client/alerting/options.go [hunk 9]
```diff
@@ -26,13 +39,37 @@ func NewAlertingOptions() *Options {
...
+			}
+		}
 	}
+
+	return errs
+}
+
+func (o *Options) AddFlags(fs *pflag.FlagSet, c *Options) {
+	fs.StringVar(&o.Endpoint, "alerting-server-endpoint", c.Endpoint,
+		"alerting server endpoint for the alerting with v1 version.")
```
Comment (benjaminhuo) [inline pkg/simple/client/alerting/options.go:67]: alerting with v1 version =>
alerting v1

**Review by benjaminhuo — CHANGES_REQUESTED** [review]

**Review by benjaminhuo — APPROVED** [review]
/lgtm

File: cmd/ks-apiserver/app/options/options.go [hunk 10]
```diff
@@ -199,6 +201,14 @@ func (s *ServerRunOptions) NewAPIServer(stopCh <-chan struct{}) (*apiserver.APIS
 		apiServer.OpenpitrixClient = opClient
 	}
 
+	if s.AlertingOptions != nil && (s.AlertingOptions.PrometheusEndpoint != "" || s.AlertingOptions.ThanosRulerEndpoint != "") {
+		alertingClient, err := alerting.NewRuleClient(s.AlertingOptions)
+		if err != nil {
+			return nil, fmt.Errorf("failed to init alerting client")
```
Comment (zryfish) [inline cmd/ks-apiserver/app/options/options.go:207]: should return error `err` as well

**Review by benjaminhuo — APPROVED** [review]
/lgtm

**Comment by zryfish** [comment]
/approve

Please verify as soon as possible

**Pull request merged**